src/aleph/http.clj

```diff
@@ -389,6 +392,9 @@
                                   ;; function.
                                   (reset! dispose-conn! (fn [] (flow/dispose pool k conn)))
 
+                                  ;; allow cancellation during connection establishment
+                                  (d/connect result (first conn))
```

Comment:
I assume the way this works is that, if result is already set, it will preempt the connection?

Can you expand the comment to explain this better?

Reply:
Indeed, that's how that works, learned that trick from https://github.com/clj-commons/manifold/issues/167#issuecomment-457996116 - and your reaction confirms Oleksii's statement in the comment after that:

> The approach is not that obvious. Even having some experience with the library.

:sweat_smile: There's another occurrence of this idiom further down this function and I already introduced another instance with https://github.com/clj-commons/aleph/pull/714. Perhaps a util function like `preempt` (which would essentially just be an alias for `d/connect`) with a corresponding docstring would do the trick best?

Also, I just noticed that users could cause some havoc by manually resolving a result promise as successful at the moment. I guess we could say this is just improper usage but ideally we'd only propagate errors in this direction :thinking: Another case in point for a util function! Alright, I convinced myself to give this a shot!

Reply:
Ah nevermind, this is in fact the only place where I used `d/connect` to propagate an error backwards like that. I will just convert it to  `d/on-realized`, too, which hopefully makes the intent more obvious. But perhaps a util for the `d/on-realized`-based error propagation is still reasonable. Will experiment and report back.

Reply:
I don't think we should worry about users messing with the deferreds they're given. Technically they could do that on _any_ unresolved deferred...but they won't, not if they want their results. :D

Reply:
One thing is still unclear; what is the advantage to using `connect` in this way over using `alt`?

Reply:
I suppose you mean instead of `(d/connect conn result)` and then returning `result`, why not return `(d/alt conn result)`  instead? If so, the problem with that is that `d/alt` will return a *new* deferred. Putting e.g. a `d/timeout!` on that one will have no bearing on `conn` then, i.e. putting the returned deferred into an error state will not propagate back  to cancel the connection establishment. Does that make sense?

Reply:
Ahh, right, right, what was I thinking? nm